# Class Activity-2

https://github.com/hasanrazaali/KNN


### Step 1: Import necessary libraries

In [54]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

### Step 2: Load the dataset

In [30]:
df = pd.read_csv("customer_data.csv")

### Step 3: Data Preprocessing

In [31]:
df.head()

,id,gender,age,salary,buy
0,15624510,Male,19.0,19000.0,0
1,15810944,Male,35.0,20000.0,0
2,15668575,Female,26.0,43000.0,0
3,15603246,Female,27.0,57000.0,0
4,15804002,Male,19.0,76000.0,0


In [32]:
df.shape

(400, 5)

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   id      400 non-null    int64  
 1   gender  398 non-null    object 
 2   age     398 non-null    float64
 3   salary  398 non-null    float64
 4   buy     400 non-null    int64  
dtypes: float64(2), int64(2), object(1)
memory usage: 15.8+ KB


In [34]:
df.isnull().sum()

id        0
gender    2
age       2
salary    2
buy       0
dtype: int64

In [35]:
df['gender'].mode()[0]

'Female'

In [36]:
df['age'].mean()

np.float64(37.6608040201005)

In [37]:
df['salary'].mean()

np.float64(69899.49748743718)

In [38]:
df['age']=df['age'].fillna(df['age'].mean())

In [40]:
df['salary']=df['salary'].fillna(df['salary'].mean())

In [39]:
df['gender']=df['gender'].fillna(df['gender'].mode()[0])

In [41]:
df.isnull().sum()

id        0
gender    0
age       0
salary    0
buy       0
dtype: int64

In [42]:
X = df[['gender', 'age', 'salary']]  # drop User ID
y = df['buy']

### Step 4: Split the dataset into a training set and a testing set

In [43]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)                                  

In [45]:
labelencoder = LabelEncoder()
X_train['gender'] = labelencoder.fit_transform(X_train['gender'])
X_test['gender'] = labelencoder.transform(X_test['gender'])

In [47]:
scaler = StandardScaler()

# Fit on training data
X_train = scaler.fit_transform(X_train)

# Apply same transformation on test data
X_test = scaler.transform(X_test)

### Step 5: Model

In [48]:
# Model
model = KNeighborsClassifier(n_neighbors=3)
model.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [49]:
# Prediction
y_pred = model.predict(X_test)

In [50]:
# Accuracy
print("Accuracy:", accuracy_score(y_test, y_pred))

Accuracy: 0.9125


### Step 6: Evaluate the model

In [ ]:
# Detailed classification metrics
# Evaluate
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))

Confusion Matrix:
 [[49  3]
 [ 4 24]]
Classification Report:
               precision    recall  f1-score   support

           0       0.92      0.94      0.93        52
           1       0.89      0.86      0.87        28

    accuracy                           0.91        80
   macro avg       0.91      0.90      0.90        80
weighted avg       0.91      0.91      0.91        80



### Step 7: Make predictions

In [53]:
new_data = pd.DataFrame([[1, 35, 60000]], columns=['gender','age','salary']) 
#Gender = Male (1), Age = 35, Estimated Salary = 60,000
new_data_scaled = scaler.transform(new_data)
prediction = model.predict(new_data_scaled)
print("Will purchase?", prediction[0])

Will purchase? 0


### Insights:   
With 91% accuracy, the model reliably identifies purchase patterns, though it performs slightly better for non-buyers than buyers, making it useful for business targeting strategies.  
1. Target High-Potential Customers   
Focus marketing campaigns on customers predicted as likely buyers to increase conversion rates and optimize marketing spend.  

2. Improve Engagement for Low-Probability Customers  
Design special offers or discounts for customers predicted as non-buyers to encourage purchases and expand the customer base.  